pkg/controller/sidecarset/sidecarset_processor.go

```diff
@@ -284,6 +330,99 @@ func (p *Processor) getSelectedPods(namespaces []string, selector labels.Selecto
+	latestRevision *apps.ControllerRevision, collisionCount int32, err error,
+) {
+	// get revision selector of this sidecarSet
+	hc := sidecarcontrol.NewHistoryControl()
+	selector, err := util.GetFastLabelSelector(
+		hc.GetRevisionLabelSelector(s),
+	)
+	if err != nil {
+		klog.Errorf("Failed to convert labels to selector, err %v, name %v", err, s.Name)
+		return nil, collisionCount, err
+	}
+
+	sidecarSetClone := s.DeepCopy()
+	// set controllerRevision namespace, this namespace must have very
+	// strict permissions for average users. Here use namespace of kruise-manager
+	sidecarSetClone.SetNamespace(webhookutil.GetNamespace())
+
+	if sidecarSetClone.Status.CollisionCount != nil {
+		collisionCount = *sidecarSetClone.Status.CollisionCount
+	}
+
+	// list all revisions
+	revisions, err := p.historyController.ListControllerRevisions(sidecarSetClone, selector)
+	if err != nil {
+		return nil, collisionCount, err
+	}
+
+	// sort revisions by increasing .Revision
+	history.SortControllerRevisions(revisions)
+	revisionCount := len(revisions)
+
+	// build a new revision from the current sidecarSet
+	latestRevision, err = hc.NewRevision(sidecarSetClone, hc.NextRevision(revisions), &collisionCount)
+	if err != nil {
+		return nil, collisionCount, err
+	}
+
+	// find any equivalent revisions
+	equalRevisions := history.FindEqualRevisions(revisions, latestRevision)
+	equalCount := len(equalRevisions)
+
+	if equalCount > 0 && history.EqualRevision(revisions[revisionCount-1], equalRevisions[equalCount-1]) {
+		// if the equivalent revision is immediately prior the update revision has not changed
+		// in case of no change
+		latestRevision = revisions[revisionCount-1]
+	} else if equalCount > 0 {
+		// if the equivalent revision is not immediately prior we will roll back by incrementing the
+		// Revision of the equivalent revision
+		// in case of roll back
+		latestRevision, err = p.historyController.UpdateControllerRevision(equalRevisions[equalCount-1], latestRevision.Revision)
+		if err != nil {
+			return nil, collisionCount, err
+		}
+	} else {
+		// if there is no equivalent revision we create a new one
+		// in case of the latest change
+		latestRevision, err = p.historyController.CreateControllerRevision(sidecarSetClone, latestRevision, &collisionCount)
+		if err != nil {
+			return nil, collisionCount, err
+		}
+		revisions = append(revisions, latestRevision)
+	}
+
+	// only store limited history revisions
+	if err = p.truncateHistory(revisions); err != nil {
+		klog.Errorf("Failed to truncate history for %s: %v", sidecarSetClone.Name, err)
+	}
+
+	return latestRevision, collisionCount, nil
+}
+
+func (p *Processor) truncateHistory(revisions []*apps.ControllerRevision) error {
+	limited := integer.IntMax(0, maxStoredRevisions)
+	if len(revisions) == 0 || len(revisions) <= limited {
+		return nil
+	}
+	if err := p.historyController.DeleteControllerRevision(revisions[0]); err != nil {
```

Comment:
Should not delete the ControllerRevisions that have existing pods in these revisions.

Reply:
So should we count references for each revisions?